### ISSUE TITLE
🚀 Feature: allow signing git commits from scaffolder action

### ISSUE THREAD

**Issue by drodil** [backstage/backstage#25934]
### 🔖 Feature description

For example, Github scaffolder action `github:repo:push` uses the helper function here to make a commit using isomorphic-git underneath: https://github.com/backstage/backstage/blob/master/plugins/scaffolder-node/src/actions/gitHelpers.ts#L23 

The other git related scaffolder actions use the same helper but it doesn't support signing the commit.

### 🎤 Context

Some repositories might require that the commit is signed with a GPG key and using scaffolder actions to create these commits will fail in the pre-receive hook if signing support is missing. 

### ✌️ Possible Implementation

We need a new secret value to config/scaffolder that Backstage users can insert their private GPG key to be used for signing. Additionally, all scaffolder actions using the helper need to support a new parameter that determines whether the commit should be signed or not.

Integrating this to isomorphic-git is quite straightforward (https://github.com/backstage/backstage/blob/master/plugins/scaffolder-node/src/scm/git.ts#L145)

```ts
// @ts-ignore
import { pgp } from '@isomorphic-git/pgp-plugin';

return git.commit({ fs, dir, message, author, committer, signingKey, onSign: pgp.sign });
```

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by benjdlambert** [comment]
So we would need a private key in config or something right? I don't think we can sign commits from a user right as we don't have the users private key in order to do this.

I wonder if when you create a commit using the GitHub API if it automatically signs the commit for you, but that feels like a step in the wrong direction given the effort in https://github.com/backstage/backstage/issues/22244.

Do you have any thoughts on a way forward here? :pray:

**Comment by drodil** [comment]
We did this for our internal action and added private GPG key to config. It's pretty straightforward with isomorphic-git - just requires quite a lot of changes to the existing actions.

**Comment by benjdlambert** [comment]
Yeah, but I think that's fine right? We can just add another config option and if there's config set and it's enabled in the action then it would use the `isomorphic-git` client to sign using the private key?

**Comment by drodil** [comment]
Yep, exactly 👍

**Comment by dotboris** [comment]
I'm not sure that this will work on GitHub when authenticating using a GitHub app. I see two cases here:

1. A template requests the user's OAuth credentials: https://backstage.io/docs/features/software-templates/writing-templates/#using-the-users-oauth-token
1. When the backstage GitHub integration is done through a GitHub app

The core of the issue is that when you're dealing with a GitHub app, you simply can't sign commits with PGP. There's no place in the admin UI or in the API where you can upload your PGP public key.

To sign commits using a GitHub app, you need to create the commits through the GitHub API:

> Signature verification for bots will only work if the request is verified and authenticated as the GitHub App or bot and contains no custom author information, custom committer information, and no custom signature information, such as Commits API.

See https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#signature-verification-for-bots

**Comment by drodil** [comment]
@dotboris thanks for your insights! That's a bummer, I didn't know that about Github apps; for other SCMs this is a no-brainer but apparently Github wants to make everything a bit harder..

Did you check out the PR for this? Would it make sense to use the API for Github apps to sign, or just drop the support for now?

**Comment by dotboris** [comment]
I've finally gotten back into commit signatures. Sorry about the radio silence.

> Did you check out the PR for this? Would it make sense to use the API for Github apps to sign, or just drop the support for now?

I haven't had time to check the PR sorry. I think that using the GitHub API to create the commit is likely a good option. I believe that the `publish:github:pull-request` action already does this.

I've done some more digging on this and this and things are more nuanced than I initially thought. 

First, I originally said that creating commits through the GitHub API was required but according to https://stackoverflow.com/a/74071223, you should be able to use `git` and to have GitHub sign commits from an app if you make the author information just right. I haven't tested this myself so I'm not sure how true that is.

Second, during some of my testing, I noticed that when a create a commit through `publish:github:pull-request` I can get the signature to work if:
- I remove all author information from the template and the `scaffolder.defaultAuthor` in the app config.
- I don't specify a `token`. This defaults to using the GitHub app's credentials

What's interesting is that if instead of using the GitHub app credentials I use the current user's credentials [through OAuth](https://backstage.io/docs/features/software-templates/writing-templates#using-the-users-oauth-token) then the commit is no longer signed. I'm not sure if I need to do something special to the author information or if OAuth apps don't have this kind of automatic commit signing.

**Comment by drodil** [comment]
Hi, @dotboris, and thanks for this! I did some testing myself and noticed you have to pass the

```yaml
forceEmptyGitAuthor: true
```

to the action to make the signing work. Let this work as a reference for those finding this information :)

**Comment by drodil** [comment]
@chromey yes and this PR had working system for Bitbucket. Unfortunately the GitHub signing is not possible through git but the GH app uses REST API for commits as stated in the conversation above. 

So it might be possible to revive this PR but remove the GitHub part. Not 100% sure of other SCM systems that were not tested.

**Comment by drodil** [comment]
@chromey I created a new PR #28816 without GitHub options. @benjdlambert could we reopen this issue and maybe get a review for the PR? Thanks!

**Comment by dotboris** [comment]
About #28816 , I just wanted to chime in to say that signing commits through PGP should work on GitHub if you use the personal access token of a regular users. You need to use the API when you're using a GitHub app. So the PGP solution for GitHub can work in some cases but not all of them.

**Comment by sysmatrix1** [comment]
We are unable to push commits to Github in our BackStage template after security team enforced git commit signing. Please let us know if the issue will be fixed for Github app.

**Comment by leogallardocw** [comment]
> Hi, [@dotboris](https://github.com/dotboris), and thanks for this! I did some testing myself and noticed you have to pass the
> 
> forceEmptyGitAuthor: true
> to the action to make the signing work. Let this work as a reference for those finding this information :)

To help more people reaching this thread:
With #31144, we can create GitHub repositories with the auto_init option (meaning that the repository is created and the main branch has a README file by default).
After that, open a pull request to your repository using dotboris solution `forceEmptyGitAuthor: true`, and the commit from by the Github APP on the PR will be signed.

This way, repos can be created from templates and be fully signed even if the first commit on main requires signing.
